cmd/registry/core/lint-openapi.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	"github.com/apigee/registry/rpc"
+	linter "github.com/googleapis/gnostic/metrics/lint"
+	"google.golang.org/protobuf/proto"
+	yaml "gopkg.in/yaml.v3"
+)
+
+// NewLintFromOpenAPI runs the API linter and returns the results.
+func NewLintFromOpenAPI(name string, b []byte) (*rpc.Lint, error) {
+	// create a tmp directory
+	root, err := ioutil.TempDir("", "registry-openapi-")
+	if err != nil {
+		return nil, err
+	}
+	name = filepath.Base(name)
+	// whenever we finish, delete the tmp directory
+	defer os.RemoveAll(root)
+	// write the file to the temp directory
+	spec, err := GUnzippedBytes(b)
+	if err != nil {
+		return nil, err
+	}
+	err = ioutil.WriteFile(filepath.Join(root, name), spec, 0644)
+	if err != nil {
+		return nil, err
+	}
+	// run the linter on the spec
+	lintFile, err := lintFileForOpenAPI(name, root)
+	if err != nil {
+		return nil, err
+	}
+	lint := &rpc.Lint{
+		Name:  name,
+		Files: []*rpc.LintFile{lintFile},
+	}
+	return lint, nil
+}
+
+func lintFileForOpenAPI(path string, root string) (*rpc.LintFile, error) {
+	cmd := exec.Command("gnostic", path, "--linter-out=.")
+	cmd.Dir = root
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+	b, err := ioutil.ReadFile(filepath.Join(root, "/linter.pb"))
+	if err != nil {
+		return nil, err
+	}
+	output := &linter.Linter{}
+	if err := proto.Unmarshal(b, output); err != nil {
+		return nil, err
+	}
+	nodeFinder, err := newNodeFinder(filepath.Join(root, path))
+	if err != nil {
+		return nil, err
+	}
+	problems := make([]*rpc.LintProblem, 0)
+	for _, message := range output.Messages {
+		problem := &rpc.LintProblem{
+			Message:    message.Message,
+			Suggestion: message.Suggestion,
+		}
+		node, err := nodeFinder.findNode(message.Keys)
+		if err == nil {
+			l := int32(node.Line)
+			c := int32(node.Column)
+			problem.Location = &rpc.LintLocation{
+				StartPosition: &rpc.LintPosition{
+					LineNumber:   l,
+					ColumnNumber: c,
+				},
+				EndPosition: &rpc.LintPosition{
+					LineNumber:   l,
+					ColumnNumber: c + int32(len(node.Value)-1),
+				},
+			}
+		}
+		problems = append(problems, problem)
+	}
+	result := &rpc.LintFile{}
+	result.Problems = problems
+	return result, err
```

Comment:
Is `err` ever non-nil now that the error check has been added above? If not, this could be `return result, nil` to remove ambiguity about the error.

Reply:
right, thanks! changed